### PR TITLE
feat(asset-index): implement complete withdrawal

### LINKED ISSUE #27: Implement `complete_withdrawal` API call
Implement `complete_withdrawal` API call detailed in the specification under AssetIndex Pallet.
## Implementation details

- Checks every pending withdrawal within PendingWithdrawals mapping
- Asserts that LockupPeriod has passed since withdrawal was initiated
- Completes unbonding process on other parachains (withdraw_unbonded)
- If successful, transfers assets to owner and delete pending withdrawal from system
- If it fails, keep withdrawal record and allow the caller to retry

### PR DESCRIPTION
**Pull request by mattsse**
## Changes

- implements the complete withdrawal extrinsic 

## Tests

```

```

## Issues

- Closes #27 
- Related #75 #76

### CODE REVIEW

**Review by clearloop — CHANGES_REQUESTED** [review]
* We need to separate these nests, really strange here LOL
* Break loop after `all_withdraw = false`
* No tests provided

File: pallets/asset-index/src/lib.rs [hunk 1]
```diff
@@ -343,18 +352,77 @@ pub mod pallet {
...
-        pub fn complete_withdraw(_origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-            todo!()
+        pub fn complete_withdraw(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+            let caller = ensure_signed(origin)?;
+
+            let current_block = frame_system::Pallet::<T>::block_number();
+            let period = T::WithdrawalPeriod::get();
+
+            PendingWithdrawals::<T>::mutate_exists(&caller, |maybe_pending| {
+                if let Some(pending) = maybe_pending.take() {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply pallets/asset-index/src/lib.rs:371]: Agreed, makes sense to issue an error if nothing to withdraw

File: pallets/asset-index/src/lib.rs [hunk 2]
```diff
@@ -343,18 +352,77 @@ pub mod pallet {
...
+                        .into_iter()
+                        .filter_map(|mut redemption| {
+                            if redemption.initiated + period > current_block {
+                                let mut all_withdrawn = true;
+                                for asset in &mut redemption.assets {
+                                    match asset.state {
+                                        RedemptionState::Initiated => {
+                                            // unbonding processes failed
+                                            // TODO retry or handle this separately?
+                                            all_withdrawn = false;
```
[the remark this answers is not in the document]
Comment (mattsse) [reply pallets/asset-index/src/lib.rs:382]: I think we should still process all assets that we can move to a `Withdrawn` state, so we actually need to iterate over all, this flag merely marks whether we can now remove the whole `Redemption` struct

**Comment by mattsse** [comment]
> * Break loop after `all_withdraw = false`
We should continue unbonding every single asset, this is merely a maker whether we can close this pending withdrawal.
> * No tests provided
Follow up on #31

**Comment by mattsse** [comment]
The are still some code paths that may change once the bonding/unbonding procedures and states are integrated with XCM

**Review by clearloop — APPROVED** [review]
One last thing is `withdraw` and `complete_withdraw` are easy to confuse, it would be better if we figure out another function name

**Pull request merged**